src/utils/multihash.rs

```diff
@@ -76,6 +76,106 @@ pub enum MultihashCode {
     Ripemd320,
 }
 
+impl MultihashCode {
+    /// Calculate the [`Multihash`] of the input byte stream.
+    pub fn digest_byte_stream<R: std::io::Read>(&self, bytes: &mut R) -> anyhow::Result<Multihash> {
+        fn hash<'a, H: Hasher, R: std::io::Read>(
+            hasher: &'a mut H,
+            bytes: &'a mut R,
+        ) -> anyhow::Result<&'a [u8]> {
+            let mut buf = [0; 1024];
+            loop {
+                let n = bytes.read(&mut buf)?;
+                if n == 0 {
+                    break;
+                }
+                if let Some(b) = buf.get(0..n) {
+                    hasher.update(b);
+                }
+            }
+            Ok(hasher.finalize())
+        }
+
+        Ok(match self {
```

Comment:
nit: do you think it's a good opportunity for a macro to avoid code duplication?

Reply:
I don't think a macro would significantly reduce the LoC, but it would hurt readability. What do you think?

Reply:
No strong position here, I don't expect this code to be updated with many new algorithms anytime soon in the future. We can leave it as is.